### PR TITLE
Apply VCS curation to `vcsProcessed` instead of to `vcs`

### LINKED ISSUE #4618: Applying package curations may lead to loosing data from VCS processed
`PackageCurationData.apply()` applies the VCS curations to `Package.vcs`, not `Package.vcsProcessed` [1]. So, the `vcsProcessed` gets set to `vcs.normalize()` [2]. So, if `vcsProcessed` was created in any non-default way, the information it represents always gets lost. For example several package managers incorporate the fallback URLs only into `vcsProcessed`, see e.g. [3]. So, a VCS curation for Maven artifacts always lead to loosing the the VCS info if it was created from the fallback URL.

**Example**

package:
```
- package:
      id: "Maven:org.aspectj:aspectjrt:1.9.6"
      ... 
       vcs:
          type: ""
          url: ""
          revision: ""
          path: ""
        vcs_processed:
          type: "Git"
          url: "https://github.com/eclipse/org.aspectj.git"
          revision: ""
          path: ""
      curations: []
```
package curation:
```
- id: "Maven:org.aspectj:aspectjrt"
  curations:
    comment: "Package resides in its own directory within repo."
    vcs:
      path: "runtime"
```
Applying above curation to above package leads to a package with:
```
       vcs:
          type: ""
          url: ""
          revision: ""
          path: "runtime"
        vcs_processed:
          type: ""
          url: ""
          revision: "runtime"
          path: ""
```

[1] https://github.com/oss-review-toolkit/ort/blob/971292669a7820e1d749dc24582e0c0aca76649d/model/src/main/kotlin/PackageCurationData.kt#L105-L116
[2] https://github.com/oss-review-toolkit/ort/blob/971292669a7820e1d749dc24582e0c0aca76649d/model/src/main/kotlin/Package.kt#L108
[3] https://github.com/oss-review-toolkit/ort/blob/971292669a7820e1d749dc24582e0c0aca76649d/analyzer/src/main/kotlin/managers/utils/MavenSupport.kt#L647-L649

### PR DESCRIPTION
**Pull request by fviernau**
When a package lacks VCS meta-data several of ORT's package manager
integrations use fallback URLs to guess the VCS location. The result of
that guess ends up in `Package.vcsProcess` but not in `Package.vcs`.

Applying a VCS package curation works by patching `Package.vcs` and
setting `Package.vcsProcessed` to `Package.vcs.normalize()`. So, the
result of the above mentioned guess based on the fallback URLs gets
discarded.

For example, applying a VCS path curation to a package with empty `vcs`
but non-empty `vcsProcessed` yields a package wich has empty VCS info
(besides the path) for both, `vcs` and `vcsProcessed`.

Avoid loosing the guessed VCS by applying the curation to `vcsProcessed`
instead of to `vcs`. Note that only the former is considered by the
downloader or scanner respectively.

Apply the normalization on top of the curation to make this change
non-breaking, even though it might make sense to drop the normalization
later on.

Fixes #4618.

### CODE REVIEW

**Comment by fviernau** [comment]
@sschuberth @mnonnenmacher @tsteenbe I'm aware we discussed the `normalize()` should be dropped. However, I had further doubts that this is a good idea. As-is vcsProcessed is always normalized, which simplifies things and I don't see the reason why to drop that simplification. E.g. the downloader would need to apply normalization before downloading and also the matching of package configurations need to adhere to the fact that URLs are not normalized anymore. I guess that also the scan storage would be affected. No, matter what we decide about dropping the normalization. I believe it should be done out of scope of this PR.

File: model/src/main/kotlin/PackageCurationData.kt [hunk 1]
```diff
@@ -108,12 +108,12 @@ private fun applyCurationToPackage(targetPackage: CuratedPackage, curation: Pack
...
-            revision = it.revision ?: base.vcs.revision,
-            path = it.path ?: base.vcs.path
-        )
-    } ?: base.vcs
+            type = it.type ?: base.vcsProcessed.type,
+            url = it.url ?: base.vcsProcessed.url,
+            revision = it.revision ?: base.vcsProcessed.revision,
+            path = it.path ?: base.vcsProcessed.path
+        ).normalize()
+    } ?: base.vcsProcessed
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/PackageCurationData.kt:116]: I think in line 132 below `vcs` now needs to stay unchanged and `vcsProcessed` needs to be set instead. This way `vcs` still represents the unmodified metadata of the package.

File: analyzer/src/funTest/assets/projects/synthetic/gradle-all-dependencies-expected-result-with-curations.yml [hunk 2]
```diff
@@ -608,7 +608,7 @@ analyzer:
             algorithm: "SHA-1"
         vcs:
           type: "Git"
-          url: "git@github.com:hamcrest/JavaHamcrest.git"
+          url: "ssh://git@github.com/hamcrest/JavaHamcrest.git"
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/assets/projects/synthetic/gradle-all-dependencies-expected-result-with-curations.yml:611]: Commit message: Typo in `Package.vcsProcess`

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

**Comment by mnonnenmacher** [comment]
> 
> 
> @sschuberth @mnonnenmacher @tsteenbe I'm aware we discussed the `normalize()` should be dropped. However, I had further doubts that this is a good idea. As-is vcsProcessed is always normalized, which simplifies things. E.g. the downloader would need to apply normalization before downloading and also the matching of package configurations need to adhere to the fact that URLs are not normalized anymore. Not sure what handling in the scan storage would be affected. Given these worries I prefer to not drop normalization.
> 
> No, matter what we decide at the end. I believe dropping normalization should be done out of scope of this PR, if at all.

IMO not normalizing curations is still the right idea but I understand that there is some migration effort involved as you describe, so I'm fine with doing it separately. To estimate the migration effort you could write a small function that goes through the curations and package configurations to check how many of them would be affected, maybe it's not that bad after all.

**Review by mnonnenmacher — APPROVED** [review]

File: model/src/main/kotlin/Package.kt [hunk 3]
```diff
@@ -103,7 +103,8 @@ data class Package(
     val vcs: VcsInfo,
 
     /**
-     * Processed VCS-related information about the [Package] that has e.g. common mistakes corrected.
+     * Processed VCS-related information about the [Package] in normalized form. The information is either derived from
+     * [vcs], guessed from additional data as fallback or empty.
```
Comment (sschuberth) [inline model/src/main/kotlin/Package.kt]: "[...] as *a* fallback, or empty."

File: model/src/main/kotlin/Package.kt [hunk 4]
```diff
@@ -104,7 +104,8 @@ data class Package(
 
     /**
      * Processed VCS-related information about the [Package] in normalized form. The information is either derived from
-     * [vcs], guessed from additional data as fallback or empty.
+     * [vcs], guessed from additional data as fallback or empty. On top of that [PackageCuration]s may have been
```
Comment (sschuberth) [inline model/src/main/kotlin/Package.kt:107]: The commit message should fix https://github.com/oss-review-toolkit/ort/issues/4618.

**Review by sschuberth — APPROVED** [review]

**Pull request merged**